src/Paprika/Utils/WorkProcessor.cs

```diff
@@ -0,0 +1,181 @@
+
+        for (int i = 0; i < threads - 1; i++)
+        {
+            ThreadPool.UnsafeQueueUserWorkItem(new WorkProcessor(data), preferLocal: false);
+        }
+
+        new WorkProcessor(data).Execute();
+
+        if (data.ActiveThreads > 0)
+        {
+            lock (data)
+            {
+                if (data.ActiveThreads > 0)
+                {
+                    // Wait for remaining to complete
+                    Monitor.Wait(data);
+                }
+            }
+        }
+    }
+
+    public static void For<TLocal>(
+        int fromInclusive,
+        int toExclusive,
+        ParallelOptions parallelOptions,
+        Func<TLocal> init,
+        Func<int, TLocal, TLocal> action,
+        Action<TLocal> @finally)
+        => InitProcessor<TLocal>.For(fromInclusive, toExclusive, parallelOptions, init, action, @finally);
+
+    private WorkProcessor(Data data)
+    {
+        _data = data;
+    }
+
+    public void Execute()
+    {
+        var i = _data.Index.GetNext();
+        while (i < _data.ToExclusive)
+        {
+            _data.Action(i);
+            i = _data.Index.GetNext();
+        }
+
+        _data.MarkThreadCompleted();
+    }
+
+    private class SharedCounter(int fromInclusive)
+    {
+        private int _index = fromInclusive;
+        public int GetNext() => Interlocked.Increment(ref _index) - 1;
+    }
+
+    private class Data(int threads, int fromInclusive, int toExclusive, Action<int> action)
+    {
+        public SharedCounter Index { get; } = new SharedCounter(fromInclusive);
+        public int ToExclusive => toExclusive;
+        public Action<int> Action => action;
+        public int ActiveThreads => threads;
+
+        public int MarkThreadCompleted()
+        {
+            var remaining = Interlocked.Decrement(ref threads);
+
+            if (remaining == 0)
+            {
+                lock (this)
+                {
+                    Monitor.Pulse(this);
+                }
+            }
+
+            return remaining;
+        }
+    }
+
+    private class InitProcessor<TLocal> : IThreadPoolWorkItem
+    {
+        private readonly Data<TLocal> _data;
+
+        public static void For(
+            int fromInclusive,
+            int toExclusive,
+            ParallelOptions parallelOptions,
+            Func<TLocal> init,
+            Func<int, TLocal, TLocal> action,
+            Action<TLocal> @finally)
+        {
+            var threads = parallelOptions.MaxDegreeOfParallelism > 0 ? parallelOptions.MaxDegreeOfParallelism : Environment.ProcessorCount;
+
+            var data = new Data<TLocal>(threads, fromInclusive, toExclusive, init, action, @finally);
+
+            for (int i = 0; i < threads - 1; i++)
+            {
+                ThreadPool.UnsafeQueueUserWorkItem(new InitProcessor<TLocal>(data), preferLocal: false);
+            }
+
+            new InitProcessor<TLocal>(data).Execute();
+
+            if (data.ActiveThreads > 0)
```

Comment:
Is it safe? What if the last thread decrements the threads, this one reads `ActiveThreads > 0` as it can be stale, and then locks on and waits without getting the pulse? Why not pulse N times or remove the check and perform active thread counting in the lock + pulse?

Reply:
The lock is a memory barrier; so it shouldn't read stale inside

Either the lock is acquired here and the inner read is `> 0` so it waits, and the pulse will be sent

Or the pulse lock is acquired first so the ActiveThreads will be read as 0, and it will skip